### PR TITLE
Feature/position bc stent

### PR DESCRIPTION
**Pull request by BenceVirtonomy**
- New boundary conditions: PositionSolidBody, PositionScaleSolidBody incl. small example and unit tests
- Restructuring with shared pointers
- SolidRelaxationStepInner : changes the zero particle position during particle relaxation
- system_resolution_ratio_  for free adaptation of particle size in ParticleAdaptation
- google test added to docker and main.yml

### CODE REVIEW

File: SPHINXsys/src/shared/particle_dynamics/relax_dynamics/relax_dynamics.h [hunk 1]
```diff
@@ -220,6 +221,40 @@ namespace SPH
...
+		protected:
+			StdLargeVec<Vecd>& pos_0_, & pos_n_, &dvel_dt_;
+			virtual void Update(size_t index_i, Real dt = 0.0) override;
+		};
+
+		/**
+		* @class SolidRelaxationStepInner
+		* @brief carry out particle relaxation step of particles within the body
+		*/
+		class SolidRelaxationStepInner : public RelaxationStepInner
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/particle_dynamics/relax_dynamics/relax_dynamics.h:244]: @Xiangyu-Hu here is the new particle relaxation, updating the pos_0_

File: SPHINXsys/src/shared/particles/particle_adaptation.cpp [hunk 2]
```diff
@@ -34,8 +35,8 @@ namespace SPH
 	{
 		sph_body_ = sph_body;
 		system_domain_bounds_ = sph_body_->getSPHSystem().system_domain_bounds_;
-		Real system_resolution = sph_body_->getSPHSystem().resolution_ref_;
-		spacing_ref_ = RefinedSpacing(system_resolution, global_refinement_level_);
+		Real body_resololution = system_resolution_ratio_ * sph_body_->getSPHSystem().resolution_ref_;
+		spacing_ref_ = RefinedSpacing(body_resololution, global_refinement_level_);
```
Comment (BenceVirtonomy) [inline SPHINXsys/src/shared/particles/particle_adaptation.cpp:39]: @Xiangyu-Hu here I added an optional scaling factor, it can be changed by SetSystemResolutionRatio

File: SPHINXsys/src/shared/particles/particle_adaptation.cpp [hunk 3]
```diff
@@ -20,6 +20,7 @@ namespace SPH
 		ParticleAdaptation(Real h_spacing_ratio, int global_refinement_level) :
 		h_spacing_ratio_(h_spacing_ratio),
 		global_refinement_level_(global_refinement_level),
+		system_resolution_ratio_(1.0),
 		local_refinement_level_(0),
 		local_coarse_level_(local_refinement_level_ / 2),
```
Comment (chuvirtonomy) [inline SPHINXsys/src/shared/particles/particle_adaptation.cpp:25]: any possible danger on this?

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/solid_dynamics.cpp [hunk 4]
```diff
@@ -273,6 +273,75 @@ namespace SPH
...
+		{
+			BoundingBox bounds = body->findBodyDomainBounds();
+			pos_0_center_ = (bounds.first + bounds.second) * 0.5;
+			translation_ = pos_end_center_ - pos_0_center_;
+		}
+		//=================================================================================================//
+		Vecd PositionSolidBody::getDisplacement(size_t index_i, Real dt)
+		{
+			// displacement from the initial position
+			Vecd pos_final = pos_0_[index_i] + translation_;
```
Comment (chuvirtonomy) [inline SPHINXsys/src/shared/particle_dynamics/solid_dynamics/solid_dynamics.cpp:292]: can index_i out of bound? any check?

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/solid_dynamics.cpp [hunk 5]
```diff
@@ -273,6 +273,75 @@ namespace SPH
...
+			Vecd displacement = (pos_final - pos_n_[index_i]) * dt / (end_time_ - GlobalStaticVariables::physical_time_);
+			return displacement;
+		}
+		//=================================================================================================//
+		void PositionSolidBody::Update(size_t index_i, Real dt)
+		{
+			// only apply in the defined time period
+			if (GlobalStaticVariables::physical_time_ >= start_time_ && GlobalStaticVariables::physical_time_ <= end_time_)
+			{
+				pos_n_[index_i] = pos_n_[index_i] + getDisplacement(index_i, dt); // displacement from the initial position
```
Comment (chuvirtonomy) [inline SPHINXsys/src/shared/particle_dynamics/solid_dynamics/solid_dynamics.cpp:302]: index_i: any check needed? out of bound?

File: SPHINXsys/src/shared/particle_dynamics/solid_dynamics/solid_dynamics.cpp [hunk 6]
```diff
@@ -273,6 +273,75 @@ namespace SPH
 			dvel_dt_ave_[index_i] = dvel_dt_[index_i];
```
Comment (chuvirtonomy) [inline SPHINXsys/src/shared/particle_dynamics/solid_dynamics/solid_dynamics.cpp:273]: index_i: any check needed? out of bound?

File: cases_high_level_simulation/test_3d_ball_position_solid_body/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+CMAKE_MINIMUM_REQUIRED(VERSION 3.10)
+set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${CMAKE_SOURCE_DIR}/cmake) # main (top) cmake dir
+set(CMAKE_VERBOSE_MAKEFILE on)
+
+STRING( REGEX REPLACE ".*/(.*)" "\\1" CURRENT_FOLDER ${CMAKE_CURRENT_SOURCE_DIR} )
+PROJECT("${CURRENT_FOLDER}")
+
+include(ImportSPHINXsysFromSource_for_3D_build)
+
+SET(LIBRARY_OUTPUT_PATH ${PROJECT_BINARY_DIR}/lib)
```
Comment (chuvirtonomy) [inline cases_high_level_simulation/test_3d_ball_position_solid_body/CMakeLists.txt:11]: please either use set or SET and consistent for all cmake operators.

**Review by chuvirtonomy — COMMENTED** [review]
Please review comments

**Review by chuvirtonomy — COMMENTED** [review]
comments

**Review by chuvirtonomy — APPROVED** [review]

**Review by chuvirtonomy — APPROVED** [review]
